### PR TITLE
Properly reflect control flow for user-defined logical operators in control flow graph.

### PR DESCRIPTION
**Pull request by AlekseyTs**

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@333fred, @mavasani, @cston, @dotnet/roslyn-compiler, @dotnet/analyzer-ioperation Please review

**Comment by AlekseyTs** [comment]
@333fred, @mavasani, @cston, @dotnet/roslyn-compiler, @dotnet/analyzer-ioperation Please review

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IBinaryOperatorExpression.cs [hunk 1]
```diff
@@ -2865,5 +2865,1435 @@ void M(bool? a, P b, P c, bool? result)
...
+            IParameterReferenceOperation: result (OperationKind.ParameterReference, Type: C) (Syntax: 'result')
+
+        IFlowCaptureOperation: 1 (OperationKind.FlowCapture, Type: null, IsImplicit) (Syntax: 'a')
+          Value: 
+            IParameterReferenceOperation: a (OperationKind.ParameterReference, Type: C) (Syntax: 'a')
+
+    Jump if False (Regular) to Block[B3]
+        IUnaryOperation (UnaryOperatorKind.True) (OperatorMethod: System.Boolean C.op_True(C? c)) (OperationKind.UnaryOperator, Type: System.Boolean, IsImplicit) (Syntax: 'a')
+          Operand: 
+            IInvalidOperation (OperationKind.Invalid, Type: C?, IsImplicit) (Syntax: 'a')
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IBinaryOperatorExpression.cs]: Do we know what bound node this corresponds to? Seems weird that there's an IInvalidOperation in compiling C# code.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IBinaryOperatorExpression.cs]: See comment above

---
In reply to: [190711693](https://github.com/dotnet/roslyn/pull/27070#discussion_r190711693) [](ancestors = 190711693)

Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IBinaryOperatorExpression.cs]: Ah, totally missed the comment, thanks.

---
In reply to: [190712265](https://github.com/dotnet/roslyn/pull/27070#discussion_r190712265) [](ancestors = 190712265,190711693)

**Comment by 333fred** [comment]
LGTM (iteration 2)

**Review by 333fred — APPROVED** [review]
LGTM (Iteration 2)

**Comment by AlekseyTs** [comment]
@mavasani, @dotnet/analyzer-ioperation Please review

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 2]
```diff
@@ -1706,6 +1712,82 @@ IOperation negateNullable(IOperation operand)
             }
         }
 
+        private IOperation VisitUserDefinedBinaryConditionalOperator(IBinaryOperation binOp, int? captureIdForResult)
+        {
+            SpillEvalStack();
+
+            Compilation compilation = ((Operation)binOp).SemanticModel.Compilation;
```
Comment (mavasani) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Probably worth saving into a field, as we are using it for multiple nodes. Can be done in a separate PR.

Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: > Probably worth saving into a field, as we are using it for multiple nodes. Can be done in a separate PR.

There is a PROTOTYPE comment about this at the beginning of the file

---
In reply to: [190737665](https://github.com/dotnet/roslyn/pull/27070#discussion_r190737665) [](ancestors = 190737665)

File: src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs [hunk 3]
```diff
@@ -1057,6 +1057,7 @@ public override void VisitBinaryOperator(IBinaryOperation operation)
 
             LogString($" ({kindStr})");
             LogHasOperatorMethodExpressionCommon(operation.OperatorMethod);
+            var unaryOperatorMethod = ((BaseBinaryOperatorExpression)operation).UnaryOperatorMethod;
```
Comment (mavasani) [inline src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs]: Shouldn't we log this if non-null? Again, can be in a separate PR.

**Review by mavasani — APPROVED** [review]

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs (hunk 2 above)
Comment (mavasani) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Ah, thanks.

File: src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs (hunk 3 above)
Comment (AlekseyTs) [reply src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs]: > Shouldn't we log this if non-null? Again, can be in a separate PR.

This is not a part of the public API. Changing the output here would cause base-line changes in a lot of tests. There is a PROTOTYPE comment about making this public, then we can deal with this part too. For now, I am comfortable with the coverage for CFG

---
In reply to: [190738488](https://github.com/dotnet/roslyn/pull/27070#discussion_r190738488) [](ancestors = 190738488)
